File: w04-associations-and-auth/d3-drills/solutions.js
function binarySearchIterative(array, target) {
  // declare variables for high, low, and mid points of the array
  var mid, low = 0,
      high = array.length - 1; // declaring multiple variables at once

  // while the low is less than the high
  while (low <= high) {
    // set mid equal to halfway between the low and the high
    mid = Math.floor((low + high) / 2);

    // if the value of indexing the array using mid is greater than the target
    if (array[mid] > target) {
      // set the high to mid - 1, start loop over
      high = mid - 1;
    // else if the value of indexing the array using mid is greater than the target
    } else if (array[mid] < target) {
      // set the low to mid - 1, start loop over
      low = mid + 1;
    } else {
      // return the midpoint when the indexed midpoint matches the target
      return mid;
    }
  }
  // return null if the target is not found
  return null;
}

var testArray = [1,2,3,4,5,6,7,8,9,10];
console.log(binarySearchIterative(testArray, 3, 1, 10));
console.log(binarySearchIterative(testArray, 7, 1, 10));

function binarySearchRecursive(array, target, low, high) {
  // break if the high is less than the low
  if (high < low) { return null; }
  // set mid to be between low and high
  var mid = Math.floor((low + high) / 2);
  // use mid to index the array and compare its value with the target
  if (array[mid] > target) {
    // if the indexed value is higher than the target,
    // call the function again with these new values
    return binarySearchRecursive(array, target, low, mid - 1);
  }
  if (array[mid] < target) {
    // if the indexed value is lower than the target,
    // adjust the midpoint and call the function again
    return binarySearchRecursive(array, target, mid + 1, high);
  }
  // return the index of the array where the target value is located
  return mid;
}

var testArray = [1,2,3,4,5,6,7,8,9,10];
console.log(binarySearchRecursive(testArray, 3, 1, 10));
console.log(binarySearchRecursive(testArray, 7, 1, 10));

File: w06-ruby-on-rails/d3-drills/solutions.rb
# input: target value, sorted array of integers

# output:
# return true or index of target if target is found
# return false if target is not present

# psuedocode
# find the low, mid, and high index of the given array
# return the mid index if the target is equal to the index of the mid array
#=> Start looping while the low index is less than the high index
# if the target is greater than value of the array's mid index
# => set the low index to the mid index
# => set the mid index to the sum of the high index minus the low index divided by two
# elsif
# => set the high index to the mid index
# => set the mid index to the sum of the high index minus the low index divided by two

def binary_search_iterative(array, target)
# declare variables for low and high positions
low_index = 0
high_index = array.length - 1
mid_index = (high_index + low_index) / 2

# while the low is less than the high
while low_index <= high_index do

    return mid_index if target == array[mid_index]

    puts "#{low_index} #{mid_index} #{high_index}"
    sleep(1.0)

    if target > array[mid_index]
      # move lower bound up to mid, recalculate new mid
      low_index = mid_index
      # set the high halfway between
      mid_index = (low_index + high_index) / 2
    elsif target < array[mid_index]
      # move upper bound to mid, recalculate new mid
      high_index = mid_index
      mid_index = (low_index + high_index) / 2
    end
  end
end

# search for valid target below mid
p binary_search_iterative([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3)
#=> expect to return index 2

def binary_search_recursive(array, target, low = 0, high = array.length - 1)
  return "#{target} value could not be found" if  low > high
  mid = (low + high) / 2
  return mid if array[mid] == target
  p "#{low} #{mid} #{high}"
  if array[mid] > target
    high = mid - 1
  else
    low = mid + 1
  end
  binary_search_recursive(array, target, low, high)
end

target = 23
numbers = [4, 8, 15, 16, 23, 42]

p binary_search_recursive(numbers, target)
#=> 4


sample_array = [1, 2, 5, 6, 8, 10, 22, 33, 37]
p binary_search_iterative(sample_array, 2)
p binary_search_recursive(sample_array, 2)

even_array = [1, 2, 5, 6, 8, 10, 22, 33, 37, 40]
p binary_search_iterative(even_array, 37)
p binary_search_recursive(even_array, 37)
